misc/python/materialize/scratch.py

```diff
@@ -0,0 +1,299 @@
+    head_rev = git.rev_parse("HEAD")
+    ssh.runv(
+        ["git", "-C", "/home/ubuntu/materialize", "config", "core.bare", "false"],
+        "ubuntu",
+        i.public_ip_address,
+    )
+    ssh.runv(
+        ["git", "-C", "/home/ubuntu/materialize", "checkout", head_rev],
+        "ubuntu",
+        i.public_ip_address,
+    )
+
+
+class MachineDesc(NamedTuple):
+    name: str
+    launch_script: Optional[str]
+    instance_type: str
+    ami: str
+    tags: Dict[str, str]
+    size_gb: int
+
+
+async def setup_all(instances: List[Any], subnet_id: str, local_pub_key: str) -> None:
+    await asyncio.gather(*(setup(i, subnet_id, local_pub_key) for i in instances))
+
+
+def launch_cluster(
+    descs: List[MachineDesc],
+    nonce: str,
+    subnet_id: str,
+    key_name: Optional[str],
+    security_group_id: str,
+    instance_profile: Optional[str],
+    extra_tags: Dict[str, str],
+) -> List[Any]:
+    """Launch a cluster of instances with a given nonce"""
+    instances = [
+        launch(
+            key_name=key_name,
+            instance_type=d.instance_type,
+            ami=d.ami,
+            tags={**d.tags, **extra_tags},
+            display_name=f"{nonce}-{d.name}",
+            size_gb=d.size_gb,
+            subnet_id=subnet_id,
+            security_group_id=security_group_id,
+            instance_profile=instance_profile,
+            nonce=nonce,
+        )
+        for d in descs
+    ]
+
+    with open(f"{os.environ['HOME']}/.ssh/id_rsa.pub") as pk:
+        local_pub_key = pk.read().strip()
+
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(setup_all(instances, subnet_id, local_pub_key))
+    loop.close()
+
+    hosts_str = "".join(
+        (f"{i.private_ip_address}\t{d.name}\n" for (i, d) in zip(instances, descs))
+    )
+
+    for i in instances:
+        ssh.runv(
+            [f"echo {shlex.quote(hosts_str)} | sudo tee -a /etc/hosts"],
+            "ubuntu",
+            i.public_ip_address,
+        )
+
+    for (i, d) in zip(instances, descs):
+        if d.launch_script:
+            ssh.runv(
+                [
+                    "cd",
+                    "~/materialize",
+                    ";",
+                    "nohup",
+                    "bash",
+                    "-c",
+                    shlex.quote(d.launch_script),
+                    ">~/mzscratch-startup.out",
+                    "2>~/mzscratch-startup.err",
+                    "&",
+                ],
+                "ubuntu",
+                i.public_ip_address,
+            )
+
+    return instances
+
+
+def get_old_instances(threshold_seconds: int) -> List[Any]:
+    def is_running(i: Any) -> bool:
+        return bool(i["State"]["Name"] == "running")
+
+    def is_old(i: Any) -> bool:
+        tags_dict = {tag["Key"]: tag["Value"] for tag in i["Tags"]}
+        created = tags_dict.get("scratch-created")
+        if created is None:
+            return False
+        created = float(created)
```

Comment:
You probably want a try/except around this. You don't want a single malformed tag to interrupt the entire cleanup process.

Reply:
I thought about this, but I think failing loudly in CI if something is wrong is better than robustness, at our stage.

Reply:
Fair enough.